### ISSUE TITLE
Potential new mini project | Children's mental health app

### ISSUE THREAD

**Issue by iteles** [dwyl/hq#488]
This is a tiny project (less than 2 sprints) as a proof of concept for a well known UK children's charity.

## Tasks
+ [x] Initial conversation (@iteles)
+ [x] Help them narrow down their MVP to something somewhat doable (@Cleop)
+ [x] Further discussion on next steps (@iteles & @Cleop)
+ [x] Mini proposal
+ [x] 'Pitch' deck
+ [x] 'Pitch' call with head of department
+ [x] Await response
+ [x] Set expectations
+ [x] Speak to interested dwylers
+ [x] Write and send contract to AfC

**Comment by iteles** [comment]
@rub1e Could you please add the budget and Case studies into the slides and make them look aesthetically pleasing? 😬

**Comment by iteles** [comment]
We had the call.
I don't think we'll get this one as there were some questions about one or two particular limitation imposed on the scope by the small budget, but this will mostly be down to how much other agencies offer to build for the same budget.

It was lovely to build a relationship with these folks in any case and we got a new slide deck out of it to re-use 😉 

I also made a lot of notes on the dwyl overview so we'll get a reusable blurb too.

**Comment by iteles** [comment]
Further questions answered last week via email, received a further email with more questions on Friday.

**Comment by iteles** [comment]
Replied and we have a few more questions. Call scheduled for Thursday.

**Comment by iteles** [comment]
Call complete. We should have a final outcome by the end of tomorrow.

If this goes ahead, here are the likely details:
+ Kick-off 2 hour workshop on Thursday 13th September, 12h00 - 14h00
+ Design & development to begin either 26th September or 1st October (ish)
+ It's a small but intense timeframe:
  + 4 days of design (basic!)
  + 16 days of development

**Comment by iteles** [comment]
This came through :tada:

Proposed team:
+ @RobStallion & @SimonLab 
+ @cleop scrum mastering
+ Design TBC
+ QA TBC

I will speak to you all on Monday about this!
😊 

@rub1e assigning to you for the contract please - let's discuss on our Tuesday standup as priority 1.

**Comment by rub1e** [comment]
Contract delivered but not yet signed, and their corresponding form has been filled in and returned.

Saved in the `private` folder in their client folder. Will add their signed contract has been returned.

**Comment by nelsonic** [comment]
# PLEASE DON'T READ THIS <br /> (_if you are easily "offended" by "difficult truths"_)!!

@rub1e was this contract ***signed***? 
The history of this issue _appears_ like it was added to "Priorities" and then not really followed through. 
_Starting_ a project without a _water-tight_ contract that includes clear IP clauses, SLAs and terms for future development is like setting off on a polar expedition without 

@iteles this project went ahead despite my protest that it would be 
"_yet another MVP that gets built without contributing any re-useable code or learning 
and ultimately goes nowhere **wasting** the Stakeholders' **time and money**_".  

+ What was the "**definition of `done`**" for the project?
+ What _metrics_ were identified as key to the _success_ of the project?
+ How are those metrics being _tracked_? (_where is the **documentation** for that?!_) 
+ Are they reviewed _daily_ by the product owner?
+ What was the desired _outcome_ for the project? 
  + Was the project's "goal" _achieved_?
  + If so, where is this _**captured**_?
    + Some of the project team shared "success" on Gitter (_private ephemeral instant chat_), but where is it _systematically_ captured so that the _stakeholders_ can determine if the project was "successful" in order to decide to further invest in it?!
  

To be _clear_ I don't want _anyone_ to think that I don't _support_ building an app that helps children record their emotions and improve their mental health. If that was the case I would have never signed up to work on Healthlocker, which while not targeted at Children, did _exactly_ this!
My _objection_ to working on this application was **_procedural_ not _conceptual_**. 

These are the things we _should_ have had _nailed_ before even _considering_ another client project:

+ [ ] A **_single unified_ workflow** that **_everyone_ follows** like a Swiss Watch. 
> I opened the "workflow consolidation" issue https://github.com/dwyl/hq/issues/434 **5 months `before`** the "Action for Children" issue was opened. In that time we should have _collectively_ sat down together to formulate what our "ideal workflow" should be and _consolidated_ it into a clear flow-chart that _anyone_ could understand and follow. 

+ [ ] A clear guiding principal for creating **_re-useable_ code** and capturing/sharing all _learning_.
> While we have built a few _good_ apps for clients in dwyl, _nobody_ has _listened_ when I beg people to not to write "throw away" code on projects where everything is re-invented each time. 
I summarised these thoughts in https://github.com/dwyl/technology-stack/issues/67 but it should have been pretty _obvious_ from the "One Metric that Matters" defined in https://github.com/dwyl/hq#one-metric-that-matters over a year ago that our . Where did the people who worked on Healthlocker capture and _share_ the learning for the project? for _example_ I documented _everything_ I did for deploying the App in: https://github.com/dwyl/learn-microsoft-azure 

+ [ ] A clear **metric** for any work we do so that we aren't just haphazardly throwing something together and hoping that it works out.
> Where is the "metrics" issue in "Action for Children" like we have for Club Soda: https://github.com/club-soda/club-soda-guide/issues/97 ? _Specifically_ what is the "OMTM" of the project and how is it tracked? (_Not "Google Analytics" that **nobody** will look at much less make any critical data-driven decisions from!_)

My _objection_ to doing this project was because it would be a ***distraction*** from the project I _did_ sign up to build: https://github.com/club-soda/club-soda-guide 
Reducing alcohol consumption and eliminating the "stigma" around drinking alcohol-free beverages is _every_ bit as _meaningful_ & important to society as building an app to help children track their emotions. The _difference_ is that kids can track their emotions using the _notes_ app on their mobile device whereas "finding alcohol free options" is a _slightly_ more challenging task. 
On 10 separate occasions people on the CS standup went off on tangents about "AfC" which were _totally irrelevant_ to CS or any of the Open Source work we were trying to do with "reuse-able code".
I _voiced_ my concerns to Inês, that the _distraction_ of "AfC" was causing _delays_ in CS, and my concern was _completely_ ignored. People kept ***hijacking*** the CS standup for "AfC" specific questions.

Ultimately, I'm _stoked_ that @iteles & co did the "AfC" without my involvement.
Because it _shows_ that "_my work here is **done**_".
You all can build client projects without my help and that's _fantastic_!
We'll ignore the fact that the project does not have a `.travis.yml` file (_i.e no Continuous Integration_) 
or that the documentation is _worse_ than a "bootcamp" MVP. 
Who _cares_ about quality?! what matters is "boshing out" another project, getting paid and moving on.
People who don't _want_ to be _excellent_ at what they do don't _care_ about _mountain_ of ***technical debt*** they are creating. Someone **`else`** will pick up the pieces.

Again, to be _clear_ I'm not trying to _detract_ from the "warm and fuzzy" feeling you get from creating an app that "children" use and are enthusiastic about. Just that you've built something that _nobody_ **`else`** can ***maintain***. Even someone _experienced_ with Elixir/Phoenix will have to "unpick" the app because the there are no standards and documentation is non-existent. The project has _barely_ finished and the ["rot"](https://en.wikipedia.org/wiki/Software_rot) has already manifest itself; Who is _responsible_ for updating the project to Phoenix 1.4? Who applies "security patches" when one of the dependencies has a flaw? Who is the registered "data controller" for safeguarding children's data?! all **`undefined`**. 

Process/workflow _matters_.
In the _real_ world, if people do things in the wrong order or "skip steps" there are _consequences_.
In an app where sensitive information can be collected, if it is not stored _encrypted_ and with appropriate safeguards and permissions it can fall into the wrong hands and people get _hurt_. 
Where is the section in "AfC" on cyberbullying? Does each user have a _password_ so they can _protect_ their content? Nope: https://github.com/actnforchildren/mental_health_app/blob/master/lib/afc/user.ex

This is turning into a "rant" which _nobody_ wants to read.
So I will just summarise with this: I'm ***delighted*** that dwyl no longer needs me to do client projects!

**Comment by iteles** [comment]
### tl;dr 
_Mea culpa_.

@nelsonic **What we proved with the Action for Children MVP was that we _do_ still need you to do client projects, _otherwise you would not have felt you had to write the comment above_.**
For the vast majority of our projects you have always been there to remind us of the things we were missing and to write tutorials on these.  

> Process/workflow _matters_. ❤️ 

We have _all_ the resources for this (or a [very](https://github.com/dwyl/learn-tdd) [large](https://github.com/dwyl/learn-travis) [number](https://github.com/dwyl/learn-istanbul) [of](https://github.com/dwyl/todomvc-vanilla-javascript-example) [them](https://github.com/dwyl/technology-stack/issues/67) thanks to you @nelsonic), but what we _clearly_ lack is a 'checklist' to be followed, furthering https://github.com/dwyl/contributing, for when _you_ can't be present. **This project has been a great case study/experiment for this** :+1:

As you've said, what we need is to take this and create the single unified process ***for building dwyl products*** (our immediate focus) as our highest priority. We could then add modifications for client projects if we were to do more after we have built out https://github.com/dwyl/technology-stack/issues/67.  
I couldn't agree more with the three checklist items you added to your comment.

I will also be delighted when dwyl no longer needs you to do client projects as that will be the true mark of a sustainable business.

**Closing this issue as the Action for Children project has been and gone and I will open the relevant issues on _process/workflow_ mentioned in @nelsonic's comment above in `hq`** to keep things separate.

<hr/>

### On the Action for Children specific questions:

Note: I have not referenced the issues below directly as the team watches Github relatively closely and I didn’t feel it was relevant for  them (or anyone else reading their public repo) to be alerted to _this_ issue by following the 'linked issues'.

This project was _very much_ about creating a `happy path prototype` to enable Action for Children (AfC) to do user testing and capture/provide undeniable metrics that will allow them to get funding to build the application into something 'real' that will effect _genuine_ change to children's mental health.

Despite having already done a lot of research into the subject, creating user personas and user testing with paper prototypes, they had ***many open questions*** around what was feasible and what _children_ would realistically use once it was an app in their hands.

![image](https://user-images.githubusercontent.com/4185328/49297293-5c9dc680-f4b2-11e8-8b5c-7200da0f224a.png)

To answer your _specific_ questions:
The contracts are in place, we have a paper trail of an agreement that the project scope has been completed (process required :woman_facepalming:) and we do not have an SLA in place because that was their specific (written) choice whilst they review it for now.


> What was the desired _outcome_ for the project?  

To improve children’s resilience in coping with change at a particularly difficult transitionary time in their lives (10-13 years of age), by using a tool that is already part of their daily lives (phones and mobile apps). Whilst this was not done as an *experiment* (there isn’t a ‘control group’), the expected change is for a tendency for an increase in ‘happy emotions’ and a decrease in ‘negative emotions’ over time.

> What was the "**definition of `done`**" for the project?  

To have an application which would allow children to perform 6 basic tasks:
+ Log their emotions on a day to day basis
+ Reflect on those emotions over time
+ Review their emotions 
+ Obtain information about coping strategies easily
+ Share those emotions with a trusted adult
+ Log in with fixed password (pin) & username
  + This application was created _as a tool_ for testing and learning with a **controlled group** of individuals who consented to a non-anonymised usage so that more could be learned from specific individuals on their behaviours, the _next_ step is to make this an application open to the public, driven by the data and funding resulting from this MVP

These features were to be delivered in such a way that the aesthetics would not detract from the UX but also not play an overpowering part.

This was defined within the GitHub project.

> What _metrics_ were identified as key to the _success_ of the project?

There was not enough _focus_ here and this was acknowledged by the PO but there was a clear set of metrics they wanted to track. There were discussions around it, but as this was a very exploratory project, the ability to capture as much data as possible on the users’ interactions with the features that were hypothesised to be useful was their chosen “focus”:

![screen shot 2018-12-01 at 14 22 51](https://user-images.githubusercontent.com/4185328/49329327-cc748580-f574-11e8-9e47-dc31366dc2f5.png)


> How are those metrics being _tracked_? (_where is the **documentation** for that?!_) 

The documentation in this project is a great example of one of our systemic issues, not having a clear checklist of what people should include in the documentation with a _good **client project** example_ https://github.com/dwyl/contributing/issues/69.

In answer to the question, they are being tracked through a combination of Hotjar (a specific request from AfC because they already use it, have an internal tech team that can make their metrics more detailed and are familiar with how to analyse that format of data) and Redash.

They have so far been reviewed weekly, but this may drop off in December after the final batch of user testing.
They picked up some unexpected results too:
![screen shot 2018-12-01 at 16 44 54](https://user-images.githubusercontent.com/4185328/49330610-de602380-f588-11e8-8653-c69e3b08ecf5.png)

> What _decisions_ have been made based on metrics so far?

This was a 3  week MVP with one set of user testing 2/3 of the way through.

_During_ this short project there were no decisions made on data, though a few were made based on what we were able to prioritise for _user testing_, like change the safeguarding or crisis information as they found that it wasn’t clear enough for children.

In the month following the project, the only decisions to have been data driven so far have been to make a couple of updates to the login screen which  they dug into when they saw that children were clicking away from this screen without logging in or were spending disproportionate amounts of time on this screen.
I have been told that they are extending their testing period so that they can monitor the analytics for longer as they think these usability issues may be skewing their data.

> Was the project's "goal" _achieved_?

According to AfC, yes; and just last week they filled in our ‘end of project’ form, but we don’t currently have a structure for how to share this publicly yet, quite a hole in our processes.
When discussing the end of the project with the scrum master, I realised I had to say ‘leave it with me, I have to write this process up’. :woman_facepalming:

They have seen continued usage from around 60% of the children and are continuing to receive useful data.  

This has been captured by the AfC team internally and they are compiling it for their funding request - it will be shared with us in due course, but has not been captured anywhere __public_ yet.

The example shared in Gitter was an example that the AfC team shared with us on a standup call - again, we need a process for sharing exactly this kind of thing :+1:


> What part of Healthlocker can we reuse for AfC?

The question was asked, but as noted in https://github.com/dwyl/technology-stack/issues/67, Healthlocker was not created in a reusable way. So whilst some of the _knowledge_ from creating HL was reused, the code was  not.

There was an effort to improve and utilise reusable components as part of the project but I had done a terrible job of explaining the full vision of https://github.com/dwyl/technology-stack/issues/67 so although `autoform` was used, the principles were a little lost for the project as a whole. https://github.com/dwyl/autoform/issues/18 https://github.com/dwyl/autoform/issues/20 https://github.com/dwyl/autoform/issues/27 https://github.com/dwyl/autoform/issues/24 were opened as part of this usage, but we had a conversation about it in our retrospective (another thing I haven’t got a process for intelligently capturing yet) and we clarified a few questions that I could have clarified during the project had I been reviewing and understanding the code for example.

https://github.com/dwyl/technology-stack/issues/67 was  a big help in clarifying a lot of the questions that I should have answered before this project began.

My conclusion from this project was two-fold:
+ For an organisation with such lengthy and wordy process documentation #434, it still has a _long_ way to go and the changes from the learnings above need to be made
+ I’m ever glad for https://github.com/dwyl/technology-stack/issues/67 as it (and clearer processes ::sweat:) will make a project like this much more seamless in the future and guarantee that whatever we do only brings positive value to the organisations we partner with

<hr/>

I suggest we close this issue as this _stage_ of the project has been done (I should have closed this much sooner, my apologies) and will capture the above in _other_ issues as actions from lessons learned.

**Comment by nelsonic** [comment]
@iteles "capturing" the points I've made in _other_ issues will simply _fragment_ points and mean that _nothing_ changes/improves. For an organisation's workflow/process/approach for delivering _excellence_ to change, the _people_ need to be willing to change/improve. 

> In the same way that the "C-level" _people_ at **Facebook** will _never_ change from harvesting people's data to sell it to advertisers who in turn _manipulate_ people into buying stuff they don't need. 

There needs to be a _desire/commitment_ to be _excellent_ at delivering the highest quality software.
That means _everyone_ needs to be _proficient_ at the chosen technology ... which has been well-defined for the past couple of years: https://github.com/dwyl/technology-stack/

I can count on _one_ hand the number of times people have _systematically captured_ their learning from _any_ of the Client projects and _shared_ it with everyone in the team and wider community.

For example, when I wanted to do field-level Encryption for TC, I _painstakingly_ wrote a _beginner's_ guide for it: https://github.com/dwyl/phoenix-ecto-encryption-example
Yes it took me "_longer_" than simply "boshing" out the code, but the _residual result_ is code that _anyone_ can ***maintain*** and not have to "_**throw away** and **start again**_". 

I _thought_ I was "_leading by example_" in that regard, but _sadly_ nobody "gets" it. 
Nobody _wants_ to _understand_ that ***only*** by systematically documenting & sharing your knowledge can you "***level up***" in _any_ organisation. 
The most "_successful_" (_both in terms of job satisfaction and financial compensation_) software engineers I know have _made_ the time to write the ***book*** on their chosen subject. 
Take Addy Osmani https://github.com/addyosmani the _reason_ he is so _incredibly_ "_successful_" is that he _systematically_ shares his knowledge: 
![image](https://user-images.githubusercontent.com/194400/49378285-1c358700-f704-11e8-9681-85a7f78babb2.png)
Each time he learns something new he writes an Open Source _book_ on the topic to both solidify his knowledge and share it with the world!

I've _tried_ to make this process _easier_ for everyone in our team(s) by making individual "bitesized" tutorials on specific topics or learning objectives, but people still drag their feet on contributing.

I'm totally _depressed_ by the number of times I've had to _cajole_ people to write better documentation for the work they do. By the number of times I've been assigned PRs to review and my comment is: 
"code looks good, please add tests and update the `README.md` ..."

I feel that _everyone_ in DWYL has life _much_ too easy building "Green Field" projects
and need to go work somewhere on _maintaining_ software for a few months/years to _understand_ the benefits of _maintainable_ code.
